osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs

```diff
@@ -234,12 +234,13 @@ private double computeDisplayStartTime(HitObjectLifetimeEntry entry)
             return algorithm.Value.GetDisplayStartTime(entry.HitObject.StartTime, startOffset, timeRange.Value, scrollLength);
         }
 
-        private void setComputedLifetimeStart(HitObjectLifetimeEntry entry)
+        private void setComputedLifetime(HitObjectLifetimeEntry entry)
         {
             double computedStartTime = computeDisplayStartTime(entry);
 
             // always load the hitobject before its first judgement offset
             entry.LifetimeStart = Math.Min(entry.HitObject.StartTime - entry.HitObject.MaximumJudgementOffset, computedStartTime);
+            entry.LifetimeEnd = entry.HitObject.GetEndTime() + timeRange.Value;
```

Comment:
I think at the very least this should take account of `entry.HitObject.MaximumJudgementOffset` like the previous line. The case here is an object scrolls past the hit point and time range but is still within the hit window.

The further concern I have, and that I would like a test for, is what happens if a hitobject scrolls from `t < LifetimeStart` to `t > LifetimeEnd` in one frame - does it get its judgement updated prior to being killed?

Reply:
I'm also uncertain the usage of `timeRange` is safe here, in particular when the overlapping algorithm is active (e.g. taiko). I get the logic of "the time range is the visible range, so waiting one full time range after the end time should be safe", but that's not quite how it works using the overlapping algorithm, since an object's visible time on screen also depends on the `MultiplierControlPoint` in that case (see something like https://github.com/ppy/osu/issues/21769, wherein a swell can be on screen for two thirds of the map).

Something more similar to how `computeDisplayStartTime()` does things may be warranted - or maybe this change can be shipped off to just catch locally or something.

Reply:
The logic I used to reason around that is that as long as the hitobject appears on the screen _once_, it will have an end time set by `DrawableHitObject.updateState()`.

However, I now realise you're absolutely correct and that is a concern.

Reply:
> I think at the very least this should take account of entry.HitObject.MaximumJudgementOffset like the previous line. The case here is an object scrolls past the hit point and time range but is still within the hit window.

I had that originally but couldn't find a scenario where it was visually required. Do you have an example I can work with?

> I'm also uncertain the usage of timeRange is safe here, in particular when the overlapping algorithm is active (e.g. taiko).

I also had concerns, as touched on in the OP but I couldn't seem to break it even when I was intentionally trying.

Reply:
> This overwrites the lifetime from the entry that this PR is setting.

The general process is that yes, the line you state overwrites the value, but is followed by a ruleset either adjusting the lifetime itself in the update state call:

https://github.com/ppy/osu/blob/7bc8908ca9c026fed1d831eb6e58df7624a8d614/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteHead.cs#L40-L41

or subsequently automatically set based on transforms applied since the line you highlighted:

https://github.com/ppy/osu/blob/3b8b56cbcbdb0365a03231aa151efb0050c12ebf/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L488-L489

The change I've made is for initial lifetime, which is what is used to decide whether to get to a DHO stage at all.

> If the above is the case then why are we adding timeRange.Value at all?

It needs to be lenient enough to ensure that if a hit object could be visible on screen, a DHO needs to be retrieved from a pool.

> What the -------- is going on with all this????? How did we produce this?????????

Reply:
> The change I've made in this PR is for initial lifetime before a DHO is assigned to a HO model. In this state, it is used to decide whether to de-pool a DHO (and potentially incur overheads) at all.

Thanks. As stupid as this entire thing is I believe this has helped me find an actual problem with this.

See patch below:

```diff
diff --git a/osu.Game.Tests/Visual/Gameplay/TestSceneDrawableScrollingRuleset.cs b/osu.Game.Tests/Visual/Gameplay/TestSceneDrawableScrollingRuleset.cs
index e4d39bb6de..b891808ecb 100644
--- a/osu.Game.Tests/Visual/Gameplay/TestSceneDrawableScrollingRuleset.cs
+++ b/osu.Game.Tests/Visual/Gameplay/TestSceneDrawableScrollingRuleset.cs
@@ -30,6 +30,7 @@
 using osuTK;
 using osuTK.Graphics;
 using JetBrains.Annotations;
+using osu.Framework.Testing;
 
 namespace osu.Game.Tests.Visual.Gameplay
 {
@@ -106,6 +107,45 @@ public void TestLifetimeRecomputedWhenTimeRangeChanges(string pooled)
             assertPosition(3, 1);
         }
 
+        [Test]
+        public void TestLifetimeEnd()
+        {
+            var beatmap = new Beatmap<TestHitObject>
+            {
+                BeatmapInfo =
+                {
+                    Difficulty = new BeatmapDifficulty
+                    {
+                        SliderMultiplier = 1
+                    },
+                    Ruleset = new OsuRuleset().RulesetInfo
+                }
+            };
+
+            for (int i = 0; i < 10; i++)
+            {
+                var h = new TestPooledHitObject();
+                h.StartTime = 100000 + i * time_range;
+                beatmap.HitObjects.Add(h);
+            }
+
+            beatmap.ControlPointInfo.Add(0, new TimingControlPoint { BeatLength = time_range / 2 });
+            beatmap.ControlPointInfo.Add(0, new EffectControlPoint { ScrollSpeed = 0.1 });
+
+            int counter = 0;
+            createTest(beatmap, d => d.RelativeScaleBeatLengthsOverride = true);
+            AddStep("hook into new result", () =>
+            {
+                var ruleset = this.ChildrenOfType<TestDrawableScrollingRuleset>().Single();
+                counter = 0;
+                ruleset.NewResult += _ => counter++;
+            });
+
+            setTime(95000);
+            setTime(999999);
+            AddAssert("all judged", () => counter, () => Is.EqualTo(10));
+        }
+
         [Test]
         public void TestRelativeBeatLengthScaleSingleTimingPoint()
         {
@@ -463,6 +503,12 @@ public DrawableTestHitObject([CanBeNull] TestHitObject hitObject)
                 });
             }
 
+            protected override void CheckForResult(bool userTriggered, double timeOffset)
+            {
+                if (timeOffset >= 0)
+                    ApplyMaxResult();
+            }
+
             protected override void Update() => LifetimeEnd = HitObject.EndTime;
         }
 

```

This test passes on `master` and fails on this branch. It's related to the concern @smoogipoo brought up before:

> The further concern I have, and that I would like a test for, is what happens if a hitobject scrolls from `t < LifetimeStart` to `t > LifetimeEnd` in one frame - does it get its judgement updated prior to being killed?

and it appears that the answer to that is "no it doesn't" with this change.

Reply:
Outside of tests, I can't seem to be able to make an example of this in gameplay. My initial concern is something like this:

<details>

```diff
diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
index 42ff1d74f3..81ff0c16bb 100644
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -40,6 +40,7 @@
 using osu.Game.Users;
 using osu.Game.Utils;
 using osuTK.Graphics;
+using osuTK.Input;

 namespace osu.Game.Screens.Play
 {
@@ -200,6 +201,13 @@ protected virtual void PrepareReplay()
             DrawableRuleset.SetRecordTarget(Score);
         }

+        protected override bool OnKeyDown(KeyDownEvent e)
+        {
+            if (e.Key == Key.Space)
+                Thread.Sleep(5000);
+            return true;
+        }
+
         [BackgroundDependencyLoader(true)]
         private void load(OsuConfigManager config, OsuGameBase game, CancellationToken cancellationToken)
         {
diff --git a/osu.Game/Screens/Play/ReplayPlayer.cs b/osu.Game/Screens/Play/ReplayPlayer.cs
index ff60dbc0d0..d00027899f 100644
--- a/osu.Game/Screens/Play/ReplayPlayer.cs
+++ b/osu.Game/Screens/Play/ReplayPlayer.cs
@@ -6,6 +6,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
@@ -115,11 +116,7 @@ public bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
                     return true;

                 case GlobalAction.TogglePauseReplay:
-                    if (GameplayClockContainer.IsPaused.Value)
-                        GameplayClockContainer.Start();
-                    else
-                        GameplayClockContainer.Stop();
-                    return true;
+                    return false;
             }

             return false;
```

</details>

With the thinking being that this could arise naturally via e.g. a system-wide or GC event that causes a stutter at just an opportune moment so that the hitobject never gets updated. But this doesn't seem to actually trigger an issue.

Also, in replays I think `FrameStabilityContainer` will ensure that we have at most 16ms elapsed between frames, and `TimeRange` is unlikely to ever go below 16ms in any case.

And the editor is whatever - we don't really care about judgements there.

Reply:
> Also, in replays I think FrameStabilityContainer will ensure that we have at most 16ms elapsed between frames, and TimeRange is unlikely to ever go below 16ms in any case.

Yeah, I went pretty deep in `LifetimeManager` logic and it's definitely not handling the scenario you brought up correctly, but frame stability is a saving grace.